### PR TITLE
[qs] remove fragments

### PR DESCRIPTION
**Pull request by zekun000**

### CODE REVIEW

**Review by sasha8 — APPROVED** [review]
As discussed offline regarding the performance degradation, previously we were computing the hashes across 10 workers (batch coordinators), and currently it is done in epoch_manager (4 workers). We can try spawning more workers to verify messages.

File: consensus/src/consensus_provider.rs
Comment (gelash) [inline consensus/src/consensus_provider.rs:64]: does this need to be a parameter?

File: consensus/src/quorum_store/types.rs
Comment (gelash) [inline consensus/src/quorum_store/types.rs:110]: This is very very important that we set it based on actual payload and not based on the batchinfo that we may have received from the client. It can be an easy mistake to make - I wonder if we could either add a test easily, or compare batch_info (at least these fields) that we compute to what we received, and ignore if they don't match?

Or assuming Batch is the thing we send to get SignedDigest now, it doesn't make sense to have num_bytes and num_txns in Batch? they will be computed, set here, and send back as a part of SignedDigest? So maybe we need two different BatchInfo's or optional fields (that we can check are None otherwise ignore).

**Review by gelash — COMMENTED** [review]
Sorry, maybe not directly relevant for this PR, but not too far off - both this comment and the other inline one are inspired by the new BatchInfo

SignedDigest used to be SignedDigest, now it is signed BatchInfo, and that's the signature we aggregate. So maybe we can adjust the name as well?

Also, it could be good to add a comment to ProofCoordinator::init_proof above the first check that ensures proof if from Self. This ensures we won't start the proof for a given digest with malicious validator's BatchInfo that may have arbitrary num_txns and num_bytes (but properly signed), destroying the aggregation of signatures to form a proof. 
However, naming is the least of the problems - the key in ProofCoordinator is digest, so a malicious validator can send proper digest back but set stupid numbers in num_txns and num_bytes. So it is critical to

File: consensus/src/quorum_store/types.rs
Comment (gelash) [reply consensus/src/quorum_store/types.rs:110]: Yes, I think this one is actually relevant to this specific PR - with fragments I don't think we were sending num_bytes and num_txns, but now we are sending BatchInfo with Batch and BatchInfo contains these numbers - but it shouldn't - because they anyway need to be recomputed by the receiver for quota and other purposes (blindly trusting would allow malicious validator to use more than quota space for instance) - currently it is done (above) and the provided values are ignored- but sending them opens up a possibility for a future bug.

**Comment by gelash** [comment]
> Sorry, maybe not directly relevant for this PR, but not too far off - both this comment and the other inline one are inspired by the new BatchInfo
> 
> SignedDigest used to be SignedDigest, now it is signed BatchInfo, and that's the signature we aggregate. So maybe we can adjust the name as well?
> 
> Also, it could be good to add a comment to ProofCoordinator::init_proof above the first check that ensures proof if from Self. This ensures we won't start the proof for a given digest with malicious validator's BatchInfo that may have arbitrary num_txns and num_bytes (but properly signed), destroying the aggregation of signatures to form a proof. However, naming is the least of the problems - the key in ProofCoordinator is digest, so a malicious validator can send proper digest back but set stupid numbers in num_txns and num_bytes. So it is critical to

Part of this is orthogonal to the given PR, so I can separately send smt to add proof coordinator init test and a comment why the first aggregated signature needs to be self.

We do have to discuss whether to rename SignedDigest though: @zekun000 @sasha8 @bchocho

File: consensus/src/quorum_store/types.rs
Comment (gelash) [reply consensus/src/quorum_store/types.rs:110]: Resolving this because Payload re-computes numbers in Batch::new constructor on the receiver side, and verify ensures that it was provided correctly. I suppose there was no need to send them to start with - but it allows re-using BatchInfo structure instead of different structs (and just adding a few lines in verify to protect against) - seems fair enough for code complexity trade-off, I probably wouldn't send, but I don't have a strong opinion

**Comment by bchocho** [comment]
I vote `SignedBatchInfo`

**Comment by bchocho** [comment]
I'm trying to understand why there are so many `TIMEOUT_BATCHES_COUNT`. I think it might be closely related to the regression you're seeing, but might not be directly from this diff.

I did notice 10 per second are from this line, which just counts whenever the messages are received, which is every tick. Can you remove it as part of this PR?
https://github.com/aptos-labs/aptos-core/blob/main/consensus/src/quorum_store/batch_generator.rs#L320

It still doesn't account for all of them, I believe.

**Comment by bchocho** [comment]
Oh, actually the entry is being removed immediately when it's ready, which means when a new signature comes in, it will be re-inited.
https://github.com/aptos-labs/aptos-core/blob/main/consensus/src/quorum_store/proof_coordinator.rs#L151

But, I now see that the `preview` cherry-pick doesn't have these issues? So maybe I'm looking at an old version

**Comment by zekun000** [comment]
@bchocho that's a good point, it should be fixed by https://github.com/aptos-labs/aptos-core/pull/7091 which is in preview branch

**Comment by gelash** [comment]
> Oh, actually the entry is being removed immediately when it's ready, which means when a new signature comes in, it will be re-inited. https://github.com/aptos-labs/aptos-core/blob/main/consensus/src/quorum_store/proof_coordinator.rs#L151
> 
> But, I now see that the `preview` cherry-pick doesn't have these issues? So maybe I'm looking at an old version

init_proof checks that the sender is self, so how can it be re-inited?

**Comment by gelash** [comment]
Might be too much, but I do see that we do a lot of value.info.xyz, where xyz is expiration, num_bytes, author, etc.
Would it make sense to add a derive rule for structs that have info: BatchInfo, to derive some BatchInfoAccessor (or a better name) trait that directly gets author(), num_bytes(), expiration() (or round() / epoch()) methods? Not sure if the macro magic is worth it for the call site simplification, you decide, just a thought.

File: consensus/src/quorum_store/batch_coordinator.rs
Comment (gelash) [inline consensus/src/quorum_store/batch_coordinator.rs:58]: should we rename to DELIVERED_BATCH_COUNT?

File: consensus/src/quorum_store/types.rs
Comment (gelash) [inline consensus/src/quorum_store/types.rs:135]: In fact, this is manually implemented for Batch struct but not for others, this is exactly the interfaces I'd make a derivable trait for all structs that include info: BatchInfo

File: consensus/src/quorum_store/batch_generator.rs
Comment (gelash) [inline consensus/src/quorum_store/batch_generator.rs:44]: @sasha8 why did we have epoch here?

File: consensus/src/quorum_store/network_listener.rs
Comment (gelash) [inline consensus/src/quorum_store/network_listener.rs:63]: this to_vec()[] looks so ugly, it was there before but anything we can do there? (can it be a method on PeerId?

**Comment by gelash** [comment]
I think I went through it, would defer to @bchocho to approve, but I can stamp too if needed :-)

File: consensus/src/quorum_store/batch_generator.rs
Comment (bchocho) [reply consensus/src/quorum_store/batch_generator.rs:44]: Looks like it's now getting from `latest_logical_time`. I'll probably revive it for the now-LogicalTime-based expirations 😆

**Review by bchocho — APPROVED** [review]
LGTM

**Pull request merged**
